File: leetcode/easy/FindSmallestLetterGreaterThanTarget.py
from typing import List

class Solution:
    def nextGreatestLetter(self, letters: List[str], target: str) -> str:
        """
        Finds the smallest character in the list that is larger than the target.
        The letters list is sorted and wraps around.

        This solution uses a manual binary search implementation.
        """
        # If the target is greater than or equal to the last element, the answer
        # must wrap around to the first element of the list.
        if target >= letters[-1]:
            return letters[0]

        # Standard binary search to find the smallest letter > target.
        left, right = 0, len(letters) - 1
        ans = letters[0]  # Default answer if target is smaller than all elements

        while left <= right:
            mid = left + (right - left) // 2
            if letters[mid] > target:
                # This is a potential answer, store it and check for a better one to the left.
                ans = letters[mid]
                right = mid - 1
            else:
                # The answer must be to the right.
                left = mid + 1
        return ans

File: leetcode/easy/FindSmallestLetterGreaterThanTarget.java
/**
 * Solution class for finding the smallest letter greater than a target.
 */
class Solution {
    /**
     * Finds the smallest character in a sorted array of characters that is strictly greater than a given target character.
     * The array is circular, meaning if the target is greater than or equal to the last character, the result is the first character in the array.
     *
     * @param letters A sorted array of characters.
     * @param target  A character to compare against.
     * @return The smallest character in the array that is greater than the target.
     */
    public char nextGreatestLetter(char[] letters, char target) {
        // The problem guarantees letters.length >= 2.
        int n = letters.length;

        // We use binary search to find the smallest character that is larger than the target.
        // The search space is [0, n]. 'high' is exclusive.
        int low = 0;
        int high = n;

        // The loop continues as long as there is a search space.
        while (low < high) {
            // Calculate the middle index to avoid potential integer overflow.
            int mid = low + (high - low) / 2;

            if (letters[mid] > target) {
                // If the middle character is greater than the target, it's a potential answer.
                // We then search in the left half (including mid) for a possibly smaller character
                // that is still greater than the target.
                high = mid;
            } else {
                // If the middle character is less than or equal to the target,
                // the answer must be in the right half.
                low = mid + 1;
            }
        }

        // After the loop, 'low' points to the index of the smallest character greater than the target.
        // If 'low' is equal to 'n', it means all characters in the array are less than or equal to the target.
        // In this case, due to the circular nature of the array, we should return the first character.
        // The modulo operator '%' elegantly handles this wrap-around case.
        // For example, if low is n, low % n will be 0, returning letters[0].
        return letters[low % n];
    }
}
